### ISSUE TITLE
Feature: Support partial config reloads

### ISSUE THREAD

**Issue by titpetric** [caddyserver/caddy#3143]
When reloading the configuration, it's usual practice to gracefully shutdown / close any existing connections after their HTTP requests have completed, and kill them off after the `grace_period`. Depending on what part of the configuration has changed, not all connections need to be closed.

When considering `nginx` and it's process model, the "workers" get re-created and new traffic gets routed to them. Caddy however is a single process which doesn't get recreated, which opens up the possibility to reload only the parts of the configurations that have changed.

Premise:

We have multiple virtual hosts defined, based on their traffic patterns:

- `www`, `img` - a combination of http and http/2 traffic,
- `api` - http, long open connections, websocket, http/2 SSE

It's common that the connections for websocket/sse are long lived, and up in the tens if not hundreds of thousands of open connections at a time. Reloading the server drops them all, which also triggers a significant network/cpu penalty for closing those sockets, and re-establishing them from the various clients. Possibly, the reconnection rate might also cause DOS issues, so we would want to keep those connections open for as long as we can, and not drop them on just any configuration reload.

Situations where this scenario of dropping websocket connections can be avoided:

- the `api` websocket server configuration has not been modified
    - modifying `www`, `img` or other virtual hosts shouldn't trigger a graceful restart of `api`
- if the `api` server has multiple location blocks, and the changes are within a particular location block
    - given routes /a, /b, /c - a change to the /a route could avoid restarting the /b and /c routes

To sum up, the ideal scenario would be:

- if the particular location block has been modified, close only those connections,
- if a particular SERVER block has been modified, only close connections to that server,
- if a particular global configuration has been modified, then gracefully close all connections

Especially in high traffic scenarios, partial reloads, e.g. per-server or server-route/location block, would be a significant benefit. It would allow us to modify a part of the complete config, without indirectly causing graceful restarts, and less graceful connection closures in other parts.

**Comment by mholt** [comment]
Thanks for the issue! A couple quick notes for when I return to this:

- Connections are opaque to the server, so what they're being used for at the application layer is unknown.
- Well-behaved websocket clients should retry if the connection is dropped.
- If we don't close connections at config changes, we need to be extremely careful to be sure that absolutely nothing about those connections has changed. I'm not sure how to do this yet.
- We could probably make grace periods configurable per-server rather than for the entire HTTP app.

If this feature is implemented, it'll probably start very conservative...

**Comment by francislavoie** [comment]
I'm not sure how viable this sounds. @mholt do you have any plans to revisit this, or do you think it would be too fundamental a change to fit with the current Caddy architecture?

**Comment by mholt** [comment]
I think it can maybe work with enough bookkeeping. I'm not sure how we can get the application to bubble down some of its semantics to lower layers though... sounds like a bad abstraction. More research is needed.

**Comment by francislavoie** [comment]
I'll close this and mark as deferred -- I think the effort to implement this will likely outweigh the benefit, but it could be later revisited if more research is done.

**Comment by crummy** [comment]
Hi, I just wanted to voice our support for this feature. Here's our use case:

* We boot up VMs on the fly for students, and allow them to connect to these VMs via a web-based VNC UI which uses websockets
* After a VM boots up, our service modifies `Caddyfile` with a new reverse proxy entry pointing to (e.g.) `/vm/<studentid> -> loclahost:<vmport>` then runs `caddy reload` 
* When a VM shuts down, the reverse happens - we remove the reverse proxy and run `caddy reload`

This all works well except that when we have VMs booting in the middle of class, every student loses their WS connection and has to reconnect. We are working on automatically reconnecting them when the connection goes down but negotiating the VNC connection seems to require a minimum of several seconds of downtime. 

For what it's worth, long ago we used nginx and did not have this issue.

**Comment by francislavoie** [comment]
@crummy disconnecting WebSocket connections is necessary to avoid memory leaks, because the connections hold a reference to the config at the time they connected. If we let them stay connected, the memory used by the previous config would not get cleaned up by Go's garbage collector.

In this case, I think it's more a situation of not having any idea how we could solve this with the current design/architecture of Caddy.

**Comment by mholt** [comment]
I wouldn't mind trying though. How do other servers do it?

**Comment by AlexMikhalev** [comment]
@crummy I have the same use case. Do you have sample config to share?

**Comment by crummy** [comment]
@AlexMikhalev the config is just a regular Caddy reverse proxy config. Something like
```
ourdomain.com {
   handle student1/* {
       reverse_proxy localhost:123
   }
   handle student2/* {
      reverse_proxy localhost:456
   }
}

**Comment by AlexMikhalev** [comment]
I figured it out via API - I am updating and removing route handles automatically: 
```
curl -X PUT -H 'Content-Type: application/json' http://localhost:2019/config/apps/http/servers/srv0/routes/0/handle/0/routes/0 -d @test3_route.json
```
The trick is to make sure it's a put and inserts a new route before "catch-all" route.

**Comment by sigaloid** [comment]
@AlexMikhalev  - does this mean you've figured out how to add routes without interrupting the existing websockets connections? I was under the impression that POST'ing to the API would trigger a config reload anyway, breaking the connections.

**Comment by francislavoie** [comment]
Any config change triggers a reload, which will cause websocket connections to close. See https://github.com/caddyserver/caddy/issues/3143#issuecomment-1384482393

**Comment by sigaloid** [comment]
Ah, okay - I misread the intention behind the comment as a possible solution. Thank you!

**Comment by francislavoie** [comment]
Worth mentioning, if the usecase is to route by dynamically update the list of proxy upstreams, or dynamically update subpath routing to proxies, then writing a custom dynamic upstreams module would allow completely avoiding reloads in those situations. See https://caddyserver.com/docs/caddyfile/directives/reverse_proxy#dynamic-upstreams

**Comment by kevinlang** [comment]
> @crummy disconnecting WebSocket connections is necessary to avoid memory leaks, because the connections hold a reference to the config at the time they connected. If we let them stay connected, the memory used by the previous config would not get cleaned up by Go's garbage collector.
> 
> In this case, I think it's more a situation of not having any idea how we could solve this with the current design/architecture of Caddy.

@francislavoie if I'm understanding correctly, once all the active connections associated with the old config drain off/close, then the memory of the old config would then be cleaned up by the garbage collector, correct? I think for the common scenario of a blue-green deployment where the Caddyfile is reloaded after it is modified to point to a new upstream (or doing a similar change via the API) this sort of memory usage is to be expected (from my perspective).

Unfortunately, the inability to maintain those connections across config reloads impacts our ability to do blue-green deployments where we slowly drain the connections to the old version of the app at the application level after switching over traffic. Are there any medium or long-term plans to support maintaining those active websocket connections across reloads? Right now we are looking at switching over to nginx as this is a necessary feature for us, but would of course prefer to stay on Caddy which is much more pleasant to work with.

Not sure if this is the right place for this discussion, let me know if I should create a forum thread or make a new issue.

**Comment by mholt** [comment]
@kevinlang To make sure I understand, why are the clients simply not reconnecting if the connection is dropped?

**Comment by kevinlang** [comment]
The clients do/can reconnect. 

Assuming we have two running applications, NEW and OLD, with Caddy currently proxying traffic to OLD. The majority of traffic being proxied are websocket connections.

What we want to do is to instruct Caddy to route all new traffic to NEW. After that, we would have our OLD application do its own graceful teardown/draining. This batched draining process would take around a minute. When OLD closes these old connections in batches, they would reconnect and establish a new connection, and since Caddy is now pointing to NEW, they would establish a connection with NEW.

However, since Caddy closes _all_ websocket connections on reload, we don't have this opportunity to do graceful draining at the application level to slowly move over connections. This also means that OLD cannot execute its own teardown logic for each connection when it receives SIGTERM, which makes the client experience more janky.

If the concern is around memory issues with keeping old websocket connections around indefinitely, perhaps there could be some configurable timeout value with a good default. After that timeout is reached, the connections are closed as they are today. This is similar in nature to [`worker_shutdown_timeout`](http://nginx.org/en/docs/ngx_core_module.html#worker_shutdown_timeout) in nature.

**Comment by mholt** [comment]
Ok, thanks. Francis, Weidi, and I talked about this in Slack. We can work on it: WebSockets can be configured to stay open through config reloads, with an optional/required(?) deadline at which they must be RST'ed. I think the deadline configured should be spread out across remaining connections to avoid thundering herd re-connections.

**Comment by francislavoie** [comment]
IMO we should add two separate new config options to `reverse_proxy` to control this:

- `stream_timeout`: a general timeout for all hijacked stream connections. This would be essentially an absolute lifetime for a stream (aka WebSocket connection)
- `stream_close_delay`: a delayed force-close of a connection after the config is being torn down

We'd need to document that `stream_close_delay` may cause connections to not close cleanly if SIGTERM is sent to Caddy, in case it's longer than the higher-level `grace_period` timeout. But it should allow a smoother transition on a config reload.

**Comment by kevinlang** [comment]
Awesome! Should we re-open this issue or create a new one? Thanks again 👍
